src/frequenz/channels/util/_timer.py

```diff
@@ -1,135 +1,664 @@
 # License: MIT
-# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
 
-"""A timer receiver that returns the timestamp every `interval`."""
+"""A timer receiver that ticks every `interval`.
 
+Note:
+    This module always use `int`s to represent time.  The time is always in
+    microseconds, so the timer resolution is 1 microsecond.
+
+    This is to avoid floating point errors when performing calculations with
+    time, which can lead to very hard to reproduce, and debug, issues.
+"""
+
+from __future__ import annotations
+
+import abc
 import asyncio
-from datetime import datetime, timedelta, timezone
-from typing import Optional
+from datetime import timedelta
 
 from .._base_classes import Receiver
 from .._exceptions import ReceiverStoppedError
 
 
-class Timer(Receiver[datetime]):
-    """A timer receiver that returns the timestamp every `interval` seconds.
+def _to_microseconds(time: float | timedelta) -> int:
+    """Convert a time or a timedelta to microseconds.
+
+    Args:
+        time: The time to convert. If it is a `float`, it is assumed to be in
+            seconds.
+
+    Returns:
+        The time in microseconds.
+    """
+    if isinstance(time, timedelta):
+        time = time.total_seconds()
+    return round(time * 1_000_000)
```

Comment:
Just a louder thought about this function. The function is overwriting the input parameter `time` and it is doing a type conversion from `timedelta` to `float` if `time` is `timedelta`. The function is correct but generally speaking I don't think this is a good practice as it can lead to error/bugs

```python
time_seconds: float
if isinstance(time, timedelta):
    time_seconds = time.total_seconds()
else:
    time_seconds = time
return round(time_seconds * 1_000_000)
```

Reply:
No, I think you are looking at this from a more conventional static typed language :)

Actually Python use "algebraic types" (I think that's how it is called). The type of `time` is `float | timedelta`, is both at the same time. But once you asserted somehow that the time is actually pinned to one of them, the type checker knows it and can still figure out the correct type. So I'm actually *narrowing* the type of `time` from `float | datetime` to `float` by converting the `datetime` to `float` when `time` is a `datetime`. Once the type is *narrowed*, the type system knows that `time` from now on is always `float`. This is why is not complaining when I'm calling `round()` (which takes a `float`) with `time`.

It would be something along these lines:

```python
    if isinstance(time, timedelta):
        time = time.total_seconds()
    assert isinstance(time, float)
    time_float: float = time
    return round(time_float * 1_000_000)
```

So to say.  Does it make sense?

Reply:
> No, I think you are looking at this from a more conventional static typed language :)

Of course :)

> So to say. Does it make sense?

It does make sense and I just need to get used to